File: apollo-api/src/main/java/com/apollographql/apollo/api/internal/SimpleOperationResponseParser.java
package com.apollographql.apollo.api.internal;

import com.apollographql.apollo.api.Error;
import com.apollographql.apollo.api.Operation;
import com.apollographql.apollo.api.Response;
import com.apollographql.apollo.api.ScalarTypeAdapters;
import org.jetbrains.annotations.NotNull;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import static com.apollographql.apollo.api.internal.Utils.checkNotNull;

public final class SimpleOperationResponseParser {
  private SimpleOperationResponseParser() {
  }

  @SuppressWarnings("unchecked")
  public static <D extends Operation.Data, W> Response<W> parse(@NotNull final Map<String, Object> response,
      @NotNull final Operation<D, W, ?> operation, @NotNull final ScalarTypeAdapters scalarTypeAdapters) {
    checkNotNull(response, "response == null");
    checkNotNull(operation, "operation == null");
    checkNotNull(scalarTypeAdapters, "scalarTypeAdapters == null");

    final D data;
    final Map<String, Object> responseData = (Map<String, Object>) response.get("data");
    if (responseData != null) {
      final SimpleResponseReader responseReader = new SimpleResponseReader(responseData, operation.variables(), scalarTypeAdapters);
      data = operation.responseFieldMapper().map(responseReader);
    } else {
      data = null;
    }

    final List<Error> errors;
    final List<Map<String, Object>> responseErrors = (List<Map<String, Object>>) response.get("errors");
    if (responseErrors != null) {
      errors = new ArrayList<>();
      for (Map<String, Object> errorPayload : responseErrors) {
        errors.add(readError(errorPayload));
      }
    } else {
      errors = null;
    }

    return Response.<W>builder(operation)
        .data(operation.wrapData(data))
        .errors(errors)
        .build();
  }

  @SuppressWarnings("unchecked")
  private static Error readError(final Map<String, Object> payload) {
    String message = null;
    final List<Error.Location> locations = new ArrayList<>();
    final Map<String, Object> customAttributes = new HashMap<>();
    for (Map.Entry<String, Object> entry : payload.entrySet()) {
      if ("message".equals(entry.getKey())) {
        Object value = entry.getValue();
        message = value != null ? value.toString() : null;
      } else if ("locations".equals(entry.getKey())) {
        List<Map<String, Object>> locationItems = (List<Map<String, Object>>) entry.getValue();
        if (locationItems != null) {
          for (Map<String, Object> item : locationItems) {
            locations.add(readErrorLocation(item));
          }
        }
      } else {
        if (entry.getValue() != null) {
          customAttributes.put(entry.getKey(), entry.getValue());
        }
      }
    }
    return new Error(message, locations, customAttributes);
  }

  private static Error.Location readErrorLocation(final Map<String, Object> data) {
    long line = -1;
    long column = -1;
    if (data != null) {
      for (Map.Entry<String, Object> entry : data.entrySet()) {
        if ("line".equals(entry.getKey())) {
          line = ((Number) entry.getValue()).longValue();
        } else if ("column".equals(entry.getKey())) {
          column = ((Number) entry.getValue()).longValue();
        }
      }
    }
    return new Error.Location(line, column);
  }
}

File: apollo-api-core/src/main/kotlin/com/apollographql/apollo/api/internal/SimpleOperationResponseParser.kt
package com.apollographql.apollo.api.internal

import com.apollographql.apollo.api.Error
import com.apollographql.apollo.api.Operation
import com.apollographql.apollo.api.Response
import com.apollographql.apollo.api.ScalarTypeAdapters

object SimpleOperationResponseParser {

  @JvmStatic
  @Suppress("UNCHECKED_CAST")
  fun <D : Operation.Data, W> parse(
      response: Map<String, *>,
      operation: Operation<D, W, *>,
      scalarTypeAdapters: ScalarTypeAdapters
  ): Response<W> {
    val responseData = response["data"] as? Map<String, Any>
    val data = responseData?.let {
      val responseReader = SimpleResponseReader(it, operation.variables(), scalarTypeAdapters)
      operation.responseFieldMapper().map(responseReader)
    }

    val responseErrors = response["errors"] as? List<Map<String, Any>>
    val errors = responseErrors?.let {
      it.map { error -> readError(error) }
    }

    return Response.builder<W>(operation)
        .data(operation.wrapData(data!!))
        .errors(errors)
        .build()
  }

  @Suppress("UNCHECKED_CAST")
  private fun readError(payload: Map<String, Any?>): Error {
    var message: String? = null
    var locations = emptyList<Error.Location>()
    val customAttributes = mutableMapOf<String, Any?>()
    for ((key, value) in payload) {
      when (key) {
        "message" -> message = value?.toString()
        "locations" -> {
          val locationItems = value as? List<Map<String, Any>>
          locations = locationItems?.map { readErrorLocation(it) } ?: emptyList()
        }
        else -> customAttributes[key] = value
      }
    }
    return Error(message, locations, customAttributes)
  }

  private fun readErrorLocation(data: Map<String, Any?>?): Error.Location {
    var line: Long = -1
    var column: Long = -1
    if (data != null) {
      for ((key, value) in data) {
        when (key) {
          "line" -> line = (value as Number).toLong()
          "column" -> column = (value as Number).toLong()
        }
      }
    }
    return Error.Location(line, column)
  }
}
